crates/transport/src/layers/cache.rs

```diff
@@ -0,0 +1,228 @@
+use crate::{TransportError, TransportErrorKind, TransportFut, TransportResult};
+use alloy_json_rpc::{
+    Id, RequestPacket, Response, ResponsePacket, ResponsePayload, RpcError, SerializedRequest,
+};
+use alloy_primitives::B256;
+use lru::LruCache;
+use parking_lot::RwLock;
+use serde::{Deserialize, Serialize};
+use serde_json::value::RawValue;
+use std::{
+    io::BufReader,
+    num::NonZeroUsize,
+    path::PathBuf,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+use tracing::trace;
+/// Cache Layer
+#[derive(Debug, Clone)]
+pub struct CacheLayer {
+    /// Config for the cache layer.
+    config: CacheConfig,
+}
+
+impl CacheLayer {
+    /// Instantiate a new cache layer with the the maximum number of
+    /// items to store.
+    #[inline]
+    pub const fn new(max_items: usize, path: PathBuf) -> Self {
+        Self { config: CacheConfig { max_items, path } }
+    }
+
+    /// Returns the maximum number of items that can be stored in the cache, set at initialization.
+    #[inline]
+    pub const fn max_items(&self) -> usize {
+        self.config.max_items
+    }
+}
+
+/// Configuration for the cache layer.
+/// For future extensibility of the configurations.
+#[derive(Debug, Clone)]
+pub struct CacheConfig {
+    /// Maximum number of items to store in the cache.
+    pub max_items: usize,
+    /// Path of the cache file.
+    pub path: PathBuf,
+}
+
+impl<S> Layer<S> for CacheLayer {
+    type Service = CachingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CachingService::new(inner, self.config.clone())
```

Comment:
this will create a new cache per service

I assume we'd like to reuse the cache here? so we'd need to lift the cache to the layer?

Reply:
Yeah moving it.